# Day 26 — Array Operations
 
**Week 5: NumPy — Numerical Computing Foundation**
 
Topics covered: element-wise arithmetic, universal functions (ufuncs), aggregation (`sum`, `min`, `max`, `mean`, `std`, `var`, `cumsum`, `cumprod`), axis-based operations, comparison operators, boolean arrays, `where`, `clip`, `abs`, `round`, `floor`, `ceil`, `sort`, `argsort`.
 
---

### Q1. Element-Wise Arithmetic
Create two arrays `a = [2, 4, 6, 8]` and `b = [1, 2, 3, 4]`. Perform and print: `a + b`, `a - b`, `a * b`, `a / b`, `a // b`, `a % b`, `a ** b`. Confirm each operation is element-wise, not Python-list concatenation.

In [1]:
# Q1: Element-wise arithmetic on two NumPy arrays

import numpy as np

a = np.array([2, 4, 6, 8])
b = np.array([1, 2, 3, 4])

print("a       :", a)
print("b       :", b)
print()
print("a + b   :", a + b)   # addition
print("a - b   :", a - b)   # subtraction
print("a * b   :", a * b)   # multiplication
print("a / b   :", a / b)   # true division (float result)
print("a // b  :", a // b)  # floor division (integer result)
print("a % b   :", a % b)   # modulo (remainder)
print("a ** b  :", a ** b)  # exponentiation

a       : [2 4 6 8]
b       : [1 2 3 4]

a + b   : [ 3  6  9 12]
a - b   : [1 2 3 4]
a * b   : [ 2  8 18 32]
a / b   : [2. 2. 2. 2.]
a // b  : [2 2 2 2]
a % b   : [0 0 0 0]
a ** b  : [   2   16  216 4096]


### How it works

- `np.array()` creates a NumPy array — this is NOT a Python list.
- With Python lists, `+` would concatenate: `[2,4] + [1,2]` → `[2,4,1,2]`. NumPy adds element by element instead.
- Each operation pairs elements at the same position: `a[0]` with `b[0]`, `a[1]` with `b[1]`, and so on.
- `a / b` returns floats (`[2. 2. 2. 2.]`). Use `//` when you want integer (floor) division.
- `a ** b` raises each element to the power of the matching element: `2**1`, `4**2`, `6**3`, `8**4`.

**Key takeaway:** NumPy arrays apply arithmetic element-by-element automatically — no loops needed.

### Q2. Scalar Operations
Create an array of integers from 1 to 10. Apply each of the following and print the result: add 5, subtract 3, multiply by 2, divide by 4, raise to the power of 2, take modulo 3. Confirm operations apply to every element simultaneously.

In [2]:
# Q2: Scalar operations applied to every element of an array

import numpy as np

arr = np.arange(1, 11)  # array([1, 2, 3, ..., 10])
print("Original :", arr)
print()

print("Add 5    :", arr + 5)
print("Sub 3    :", arr - 3)
print("Mul 2    :", arr * 2)
print("Div 4    :", arr / 4)   # returns floats
print("Power 2  :", arr ** 2)
print("Mod 3    :", arr % 3)

Original : [ 1  2  3  4  5  6  7  8  9 10]

Add 5    : [ 6  7  8  9 10 11 12 13 14 15]
Sub 3    : [-2 -1  0  1  2  3  4  5  6  7]
Mul 2    : [ 2  4  6  8 10 12 14 16 18 20]
Div 4    : [0.25 0.5  0.75 1.   1.25 1.5  1.75 2.   2.25 2.5 ]
Power 2  : [  1   4   9  16  25  36  49  64  81 100]
Mod 3    : [1 2 0 1 2 0 1 2 0 1]


### How it works

- `np.arange(1, 11)` creates integers from 1 up to (but not including) 11 — same idea as `range()`.
- Writing `arr + 5` applies `+5` to **every element** at once. This is called **broadcasting** (spreading a scalar across an array).
- No loop is needed — NumPy handles it internally, and it is much faster than a Python `for` loop.
- `arr / 4` returns floats (e.g. `0.25`, `0.5`). All other operations here keep integers.
- Think of it like adjusting the brightness of every pixel in a photo at once — one instruction, all elements change.

**Key takeaway:** A single scalar operation on a NumPy array applies to every element simultaneously — this is broadcasting at its simplest form.

### Q3. Universal Functions (ufuncs) — Math
Given an array of values `[0, 30, 45, 60, 90]` (degrees), convert them to radians first, then compute `np.sin`, `np.cos`, and `np.tan`. Also compute `np.exp`, `np.log`, `np.log2`, `np.log10` on a separate array `[1, np.e, 10, 100]`. Print results rounded to 4 decimal places.

In [3]:
# Q3: Universal functions (ufuncs) — trig and logarithmic operations

import numpy as np

# --- Trigonometry ---
degrees = np.array([0, 30, 45, 60, 90])
radians = np.deg2rad(degrees)          # convert degrees → radians first

print("Degrees :", degrees)
print("Radians :", np.round(radians, 4))
print()
print("sin     :", np.round(np.sin(radians), 4))
print("cos     :", np.round(np.cos(radians), 4))
print("tan     :", np.round(np.tan(radians), 4))  # tan(90°) → very large number (≈inf)

print()

# --- Exponential and Logarithms ---
vals = np.array([1, np.e, 10, 100])   # np.e ≈ 2.718
print("Values  :", np.round(vals, 4))
print()
print("exp     :", np.round(np.exp(vals), 4))     # e raised to each value
print("log     :", np.round(np.log(vals), 4))     # natural log (base e)
print("log2    :", np.round(np.log2(vals), 4))    # log base 2
print("log10   :", np.round(np.log10(vals), 4))   # log base 10

Degrees : [ 0 30 45 60 90]
Radians : [0.     0.5236 0.7854 1.0472 1.5708]

sin     : [0.     0.5    0.7071 0.866  1.    ]
cos     : [1.     0.866  0.7071 0.5    0.    ]
tan     : [0.00000000e+00 5.77400000e-01 1.00000000e+00 1.73210000e+00
 1.63312394e+16]

Values  : [  1.       2.7183  10.     100.    ]

exp     : [2.71830000e+00 1.51543000e+01 2.20264658e+04 2.68811714e+43]
log     : [0.     1.     2.3026 4.6052]
log2    : [0.     1.4427 3.3219 6.6439]
log10   : [0.     0.4343 1.     2.    ]


### How it works

- NumPy trig functions expect **radians**, not degrees. `np.deg2rad()` converts for you: 90° → π/2 ≈ 1.5708.
- `np.sin`, `np.cos`, `np.tan` are **ufuncs** — universal functions that apply to every element at once, just like `+` or `*`.
- `np.tan(90°)` produces a very large number (not true infinity) because computers cannot represent π/2 exactly.
- `np.exp(x)` computes eˣ. `np.log(x)` is the inverse — natural log base *e*.
- `np.log(np.e)` → `1.0` and `np.log10(100)` → `2.0` are good sanity checks to confirm correctness.
- `np.round(array, 4)` rounds every element to 4 decimal places in one call.

**Key takeaway:** Ufuncs are NumPy's built-in math functions — they operate element-wise on entire arrays with no loops needed.

### Q4. ufuncs — `np.sqrt`, `np.abs`, `np.sign`
Create an array `[-9, -4, -1, 0, 1, 4, 9]`. Compute `np.abs`, `np.sign`, and `np.sqrt(np.abs(...))`. Show that `np.sqrt` on negative numbers produces `nan` with a warning, and handle this gracefully.

In [4]:
# Q4: ufuncs — np.sqrt, np.abs, np.sign, and handling negatives gracefully

import numpy as np

arr = np.array([-9, -4, -1, 0, 1, 4, 9], dtype=float)  # float needed for sqrt/nan
print("Original :", arr)
print()

# --- abs and sign ---
print("np.abs   :", np.abs(arr))       # removes negative sign from every element
print("np.sign  :", np.sign(arr))      # -1 for negative, 0 for zero, +1 for positive
print()

# --- sqrt on negatives produces nan with a RuntimeWarning ---
print("np.sqrt (raw) — expect warning:")
sqrt_raw = np.sqrt(arr)                # negatives → nan, warning printed to console
print("Result   :", sqrt_raw)
print()

# --- Graceful handling: sqrt of absolute values ---
sqrt_abs = np.sqrt(np.abs(arr))        # always safe — abs removes negatives first
print("np.sqrt(np.abs(arr)):", np.round(sqrt_abs, 4))

# --- Suppress warning explicitly using np.errstate ---
print()
print("With np.errstate (no warning printed):")
with np.errstate(invalid='ignore'):    # temporarily silence the RuntimeWarning
    sqrt_silent = np.sqrt(arr)
    sqrt_silent = np.where(np.isnan(sqrt_silent), np.nan, sqrt_silent)  # keep nans visible
print("Result   :", sqrt_silent)

Original : [-9. -4. -1.  0.  1.  4.  9.]

np.abs   : [9. 4. 1. 0. 1. 4. 9.]
np.sign  : [-1. -1. -1.  0.  1.  1.  1.]

np.sqrt (raw) — expect warning:
Result   : [nan nan nan  0.  1.  2.  3.]

np.sqrt(np.abs(arr)): [3. 2. 1. 0. 1. 2. 3.]

With np.errstate (no warning printed):
Result   : [nan nan nan  0.  1.  2.  3.]


/tmp/ipykernel_20969/3591649548.py:16: RuntimeWarning: invalid value encountered in sqrt
  sqrt_raw = np.sqrt(arr)                # negatives → nan, warning printed to console


### How it works

- `np.abs(arr)` returns the absolute value of each element — negatives become positive, others unchanged.
- `np.sign(arr)` labels each element: `-1` (negative), `0` (zero), or `+1` (positive). Useful for direction checks.
- `np.sqrt` of a negative number has no real answer — NumPy fills it with `nan` (Not a Number) and prints a `RuntimeWarning`.
- The cleanest fix is `np.sqrt(np.abs(arr))` — take the absolute value first, then square-root safely.
- `np.errstate(invalid='ignore')` is a context manager that silences the warning when you **intentionally** pass negatives and handle `nan` yourself.
- `dtype=float` is set on the array because `nan` is a float concept — integer arrays cannot hold it.

**Key takeaway:** `np.sqrt` on negatives gives `nan` with a warning; always sanitize input with `np.abs` or suppress intentionally with `np.errstate`.

### Q5. Aggregation — Global
Create a 1D array of 20 random integers between 1 and 100. Compute and print: `sum`, `min`, `max`, `mean`, `median`, `std`, `var`, `cumsum`, `cumprod`. Label each output clearly.

In [5]:
# Q5: Global aggregation functions on a 1D array of random integers

import numpy as np

rng = np.random.default_rng(seed=42)              # seed for reproducibility
arr = rng.integers(1, 101, size=20)               # 20 random ints in range [1, 100]
print("Array   :", arr)
print("Shape   :", arr.shape)
print()

print("sum     :", np.sum(arr))                   # total of all elements
print("min     :", np.min(arr))                   # smallest value
print("max     :", np.max(arr))                   # largest value
print("mean    :", np.mean(arr))                  # average
print("median  :", np.median(arr))                # middle value when sorted
print("std     :", np.round(np.std(arr), 4))      # standard deviation (spread)
print("var     :", np.round(np.var(arr), 4))      # variance (std squared)
print()

print("cumsum  :", np.cumsum(arr))                # running total at each position
print()
print("cumprod :", np.cumprod(arr))               # running product — grows very fast!

Array   : [ 9 78 66 44 44 86  9 70 21 10 53 98 74 77 72 79 52 13 84 46]
Shape   : (20,)

sum     : 1085
min     : 9
max     : 98
mean    : 54.25
median  : 59.5
std     : 28.0515
var     : 786.8875

cumsum  : [   9   87  153  197  241  327  336  406  427  437  490  588  662  739
  811  890  942  955 1039 1085]

cumprod : [                   9                  702                46332
              2038608             89698752           7714092672
          69426834048        4859878383360      102057446050560
     1020574460505600    54090446406796800  5300863747866086400
  4882291794189809664  7001586678424311808  6052150856392556544
 -1495428261436375040 -3975293299853295616  3661419323035811840
 -6035426118054182912  -928440324849139712]


### How it works

- `np.random.default_rng(seed=42)` creates a random number generator. The `seed` makes results identical every run — essential for reproducible notebooks.
- `rng.integers(1, 101, size=20)` generates integers from 1 up to (not including) 101.
- `mean` is the sum divided by count. `median` is the middle value after sorting — less affected by extreme values.
- `std` (standard deviation) measures how spread out values are. `var` is just `std²` — both describe spread.
- `cumsum` at position `i` is the sum of all elements from index `0` to `i`. Think of a running bank balance.
- `cumprod` multiplies everything together cumulatively — numbers explode very quickly for values > 1.

**Key takeaway:** NumPy aggregations collapse an entire array into a single number (or running sequence for `cumsum`/`cumprod`) in one fast call.

### Q6. Aggregation — Along Axes
Create a 4×5 array of random floats. Compute `sum`, `mean`, `max`, and `min` along:
- `axis=0` (collapse rows → result per column)
- `axis=1` (collapse columns → result per row)
Print each result with its shape and explain what the axis parameter means conceptually.

In [6]:
# Q6: Axis-based aggregation on a 2D array

import numpy as np

rng = np.random.default_rng(seed=7)
arr = rng.random((4, 5))                          # shape (4, 5) — 4 rows, 5 columns
print("Array (4×5):\n", np.round(arr, 3))
print("Shape:", arr.shape)
print()

# --- axis=0: collapse rows → one result per COLUMN ---
print("=== axis=0 (collapse rows, result per column) ===")
print("sum   :", np.round(np.sum(arr,  axis=0), 3), " shape:", np.sum(arr,  axis=0).shape)
print("mean  :", np.round(np.mean(arr, axis=0), 3), " shape:", np.mean(arr, axis=0).shape)
print("max   :", np.round(np.max(arr,  axis=0), 3), " shape:", np.max(arr,  axis=0).shape)
print("min   :", np.round(np.min(arr,  axis=0), 3), " shape:", np.min(arr,  axis=0).shape)
print()

# --- axis=1: collapse columns → one result per ROW ---
print("=== axis=1 (collapse columns, result per row) ===")
print("sum   :", np.round(np.sum(arr,  axis=1), 3), " shape:", np.sum(arr,  axis=1).shape)
print("mean  :", np.round(np.mean(arr, axis=1), 3), " shape:", np.mean(arr, axis=1).shape)
print("max   :", np.round(np.max(arr,  axis=1), 3), " shape:", np.max(arr,  axis=1).shape)
print("min   :", np.round(np.min(arr,  axis=1), 3), " shape:", np.min(arr,  axis=1).shape)

Array (4×5):
 [[0.625 0.897 0.776 0.225 0.3  ]
 [0.874 0.005 0.821 0.797 0.468]
 [0.303 0.278 0.255 0.445 0.505]
 [0.553 0.996 0.793 0.622 0.989]]
Shape: (4, 5)

=== axis=0 (collapse rows, result per column) ===
sum   : [2.355 2.176 2.644 2.09  2.262]  shape: (5,)
mean  : [0.589 0.544 0.661 0.522 0.565]  shape: (5,)
max   : [0.874 0.996 0.821 0.797 0.989]  shape: (5,)
min   : [0.303 0.005 0.255 0.225 0.3  ]  shape: (5,)

=== axis=1 (collapse columns, result per row) ===
sum   : [2.823 2.965 1.786 3.953]  shape: (4,)
mean  : [0.565 0.593 0.357 0.791]  shape: (4,)
max   : [0.897 0.874 0.505 0.996]  shape: (4,)
min   : [0.225 0.005 0.255 0.553]  shape: (4,)


### How it works

- A 4×5 array has `axis=0` pointing **down the rows** and `axis=1` pointing **across the columns**.
- A simple mental trick: the axis number tells you **which dimension disappears** in the result.
  - `axis=0` removes the row dimension (4) → result shape is `(5,)` — one value per column.
  - `axis=1` removes the column dimension (5) → result shape is `(4,)` — one value per row.
- Think of a gradebook: 4 students, 5 test scores each. `axis=0` gives the class average per test. `axis=1` gives each student's average across all tests.
- `rng.random((4, 5))` generates floats uniformly between 0.0 and 1.0 — no need to scale for this exercise.

**Key takeaway:** `axis=0` collapses **across rows** (result per column); `axis=1` collapses **across columns** (result per row) — the specified axis is the one that disappears.

### Q7. `keepdims=True`
Using the array from Q6, compute `mean(axis=1, keepdims=True)`. Explain why `keepdims=True` is important when you need to subtract the row mean from every element of that row (broadcasting requires matching dimensions).
 

In [7]:
# Q7: keepdims=True and why it matters for broadcasting

import numpy as np

rng = np.random.default_rng(seed=7)
arr = rng.random((4, 5))                               # same array as Q6
print("Array (4×5):\n", np.round(arr, 3))
print("Shape:", arr.shape)
print()

# --- Without keepdims ---
mean_no_keep = np.mean(arr, axis=1)                    # shape (4,) — 1D, dimension lost
print("mean(axis=1)              :", np.round(mean_no_keep, 3), " shape:", mean_no_keep.shape)

# --- With keepdims ---
mean_keep = np.mean(arr, axis=1, keepdims=True)        # shape (4, 1) — dimension preserved
print("mean(axis=1, keepdims=True):\n", np.round(mean_keep, 3), " shape:", mean_keep.shape)
print()

# --- Why it matters: subtracting row mean from every element ---
# This FAILS without keepdims (shape mismatch: (4,5) - (4,) is ambiguous)
try:
    result_bad = arr - mean_no_keep                    # numpy may broadcast unexpectedly
    print("Without keepdims — shape of result:", result_bad.shape)  # (4,4) — WRONG!
except ValueError as e:
    print("Without keepdims — ERROR:", e)

# This WORKS correctly with keepdims (shape: (4,5) - (4,1) → (4,5))
result_good = arr - mean_keep                          # each row gets its own mean subtracted
print("With keepdims    — shape of result:", result_good.shape)     # (4,5) — correct
print("Row-mean-centered array:\n", np.round(result_good, 3))

# Verify: row means of centered array should all be ~0
print("\nRow means after centering:", np.round(np.mean(result_good, axis=1), 10))

Array (4×5):
 [[0.625 0.897 0.776 0.225 0.3  ]
 [0.874 0.005 0.821 0.797 0.468]
 [0.303 0.278 0.255 0.445 0.505]
 [0.553 0.996 0.793 0.622 0.989]]
Shape: (4, 5)

mean(axis=1)              : [0.565 0.593 0.357 0.791]  shape: (4,)
mean(axis=1, keepdims=True):
 [[0.565]
 [0.593]
 [0.357]
 [0.791]]  shape: (4, 1)

Without keepdims — ERROR: operands could not be broadcast together with shapes (4,5) (4,) 
With keepdims    — shape of result: (4, 5)
Row-mean-centered array:
 [[ 0.06   0.333  0.211 -0.339 -0.265]
 [ 0.281 -0.588  0.228  0.204 -0.125]
 [-0.054 -0.079 -0.102  0.088  0.147]
 [-0.237  0.205  0.002 -0.168  0.198]]

Row means after centering: [ 0. -0.  0.  0.]


### How it works

- Without `keepdims`, `mean(axis=1)` returns shape `(4,)` — a flat 1D array. The row dimension is gone.
- With `keepdims=True`, the result keeps shape `(4, 1)` — a column vector. The dimension stays, just shrunk to size 1.
- NumPy broadcasting rule: dimensions must be **equal or one of them must be 1** to stretch across.
  - `(4, 5) - (4, 1)` → NumPy stretches the `1` across 5 columns → correct shape `(4, 5)`.
  - `(4, 5) - (4,)` → NumPy tries to align from the right: `5 vs 4` → mismatch or wrong result.
- Subtracting the row mean from every element is called **mean-centering** — a common step before normalization in ML.
- The verification at the end confirms correctness: every row mean of the centered array equals 0.

**Key takeaway:** `keepdims=True` preserves the collapsed dimension as size 1, which lets broadcasting align shapes correctly when you need to apply a per-row or per-column result back onto the original array.

### Q8. `np.cumsum` and `np.cumprod`
Create an array `[1, 2, 3, 4, 5]`. Compute `cumsum` and `cumprod` along the default axis. Then create a 3×4 array and compute `cumsum` along axis 0 and axis 1. Explain a real-world use of cumulative sum (e.g., running total of sales).

In [8]:
# Q8: cumsum and cumprod — 1D and 2D with axis control

import numpy as np

# --- 1D cumsum and cumprod ---
arr1d = np.array([1, 2, 3, 4, 5])
print("1D array       :", arr1d)
print("cumsum         :", np.cumsum(arr1d))    # running total: 1, 3, 6, 10, 15
print("cumprod        :", np.cumprod(arr1d))   # running product: 1, 2, 6, 24, 120
print()

# --- 2D cumsum along axes ---
rng = np.random.default_rng(seed=3)
arr2d = rng.integers(1, 10, size=(3, 4))       # 3 rows, 4 columns, values 1–9
print("2D array (3×4):\n", arr2d)
print("Shape:", arr2d.shape)
print()

# axis=0: running total DOWN each column (across rows)
cs_axis0 = np.cumsum(arr2d, axis=0)
print("cumsum axis=0 (down columns):\n", cs_axis0)
print("Shape:", cs_axis0.shape)
print()

# axis=1: running total ACROSS each row (across columns)
cs_axis1 = np.cumsum(arr2d, axis=1)
print("cumsum axis=1 (across rows):\n", cs_axis1)
print("Shape:", cs_axis1.shape)
print()

# --- Real-world example: running total of daily sales ---
daily_sales = np.array([1200, 850, 1400, 960, 1100, 780, 1350])
days = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]
running_total = np.cumsum(daily_sales)

print("Daily sales  :", daily_sales)
print("Running total:", running_total)
for day, sales, total in zip(days, daily_sales, running_total):
    print(f"  {day}: sold {sales:>5} | week total so far: {total}")

1D array       : [1 2 3 4 5]
cumsum         : [ 1  3  6 10 15]
cumprod        : [  1   2   6  24 120]

2D array (3×4):
 [[8 1 2 3]
 [2 8 8 6]
 [1 1 3 4]]
Shape: (3, 4)

cumsum axis=0 (down columns):
 [[ 8  1  2  3]
 [10  9 10  9]
 [11 10 13 13]]
Shape: (3, 4)

cumsum axis=1 (across rows):
 [[ 8  9 11 14]
 [ 2 10 18 24]
 [ 1  2  5  9]]
Shape: (3, 4)

Daily sales  : [1200  850 1400  960 1100  780 1350]
Running total: [1200 2050 3450 4410 5510 6290 7640]
  Mon: sold  1200 | week total so far: 1200
  Tue: sold   850 | week total so far: 2050
  Wed: sold  1400 | week total so far: 3450
  Thu: sold   960 | week total so far: 4410
  Fri: sold  1100 | week total so far: 5510
  Sat: sold   780 | week total so far: 6290
  Sun: sold  1350 | week total so far: 7640


### How it works

- `cumsum` at index `i` = sum of all elements from index `0` to `i`. Each position accumulates the previous.
- `cumprod` does the same but multiplies — values grow extremely fast, even for small arrays.
- For a 2D array, `cumsum(axis=0)` runs the accumulation **downward** through rows, per column. `cumsum(axis=1)` runs it **rightward** through columns, per row.
- Output shape is always identical to the input shape — no dimension is collapsed (unlike `sum`).
- Real-world uses of `cumsum`:
  - Running sales totals across days or months.
  - Tracking cumulative rainfall or temperature over a season.
  - Computing a portfolio's growth curve over time.

**Key takeaway:** `cumsum` and `cumprod` return a same-shaped array where each value includes all previous values — perfect for running totals, growth tracking, and time-series analysis.

### Q9. Comparison Operators — Boolean Arrays
Create an array `a = [15, 3, 22, 8, 17, 6, 25, 11]`. Without using loops, produce boolean arrays for: `a > 10`, `a == 15`, `a != 8`, `a >= 17`. Then count how many elements satisfy each condition using `np.sum(boolean_array)`.

In [9]:
# Q9: Comparison operators producing boolean arrays — no loops needed

import numpy as np

a = np.array([15, 3, 22, 8, 17, 6, 25, 11])
print("Array:", a)
print()

# --- Boolean arrays from comparisons ---
gt10  = a > 10
eq15  = a == 15
ne8   = a != 8
ge17  = a >= 17

print("a > 10  :", gt10,  "→ count:", np.sum(gt10))   # True counts as 1
print("a == 15 :", eq15,  "→ count:", np.sum(eq15))
print("a != 8  :", ne8,   "→ count:", np.sum(ne8))
print("a >= 17 :", ge17,  "→ count:", np.sum(ge17))
print()

# --- Combining conditions with & (and) and | (or) ---
between = (a > 10) & (a < 20)                         # parentheses required around each condition
print("10 < a < 20 :", between, "→ count:", np.sum(between))

either  = (a < 5) | (a > 20)
print("a < 5 or > 20:", either, "→ count:", np.sum(either))
print()

# --- Boolean array as a mask to extract matching values ---
print("Values where a > 10   :", a[gt10])              # fancy indexing with boolean mask
print("Values where a >= 17  :", a[ge17])

Array: [15  3 22  8 17  6 25 11]

a > 10  : [ True False  True False  True False  True  True] → count: 5
a == 15 : [ True False False False False False False False] → count: 1
a != 8  : [ True  True  True False  True  True  True  True] → count: 7
a >= 17 : [False False  True False  True False  True False] → count: 3

10 < a < 20 : [ True False False False  True False False  True] → count: 3
a < 5 or > 20: [False  True  True False False False  True False] → count: 3

Values where a > 10   : [15 22 17 25 11]
Values where a >= 17  : [22 17 25]


### How it works

- Applying `>`, `==`, `!=`, `>=` to a NumPy array returns a **boolean array** — same shape, each element is `True` or `False`.
- No loop is needed — NumPy tests every element simultaneously.
- `np.sum(boolean_array)` counts `True` values because NumPy treats `True` as `1` and `False` as `0`.
- To combine two conditions use `&` (both must be true) or `|` (either must be true). Always wrap each condition in parentheses — Python's operator precedence will cause bugs without them.
- A boolean array used inside `[]` is called a **mask** — it filters and returns only the elements where the condition is `True`.

**Key takeaway:** Comparison operators on NumPy arrays return boolean arrays that can count matches with `np.sum`, combine with `&`/`|`, and filter values directly using mask indexing — all without a single loop.

### Q10. Compound Boolean Conditions
Using the array from Q9, find all elements that are: greater than 5 AND less than 20, OR equal to 25. Use `&`, `|`, and `~` operators (not Python's `and`/`or`). Explain why Python's `and`/`or` don't work on arrays.
 

In [10]:
# Q10: Compound boolean conditions using &, |, ~ on NumPy arrays

import numpy as np

a = np.array([15, 3, 22, 8, 17, 6, 25, 11])
print("Array:", a)
print()

# --- Compound condition: (5 < a < 20) OR (a == 25) ---
cond_range = (a > 5) & (a < 20)       # parentheses mandatory — & binds tighter than > and 
cond_exact = (a == 25)
combined   = cond_range | cond_exact   # either condition satisfied

print("a > 5           :", a > 5)
print("a < 20          :", a < 20)
print("(a>5) & (a<20)  :", cond_range)
print("a == 25         :", cond_exact)
print("combined (|)    :", combined)
print("Matching values :", a[combined])
print("Count           :", np.sum(combined))
print()

# --- NOT operator ~ (invert a boolean array) ---
not_combined = ~combined
print("~combined (NOT) :", not_combined)
print("Excluded values :", a[not_combined])
print()

# --- Why Python's and/or fail on arrays ---
print("Demonstrating why 'and'/'or' fail:")
try:
    result = (a > 5) and (a < 20)     # Python tries to evaluate entire array as one True/False
except ValueError as e:
    print("Error with 'and':", e)

try:
    result = (a > 5) or (a < 20)
except ValueError as e:
    print("Error with 'or' :", e)

Array: [15  3 22  8 17  6 25 11]

a > 5           : [ True False  True  True  True  True  True  True]
a < 20          : [ True  True False  True  True  True False  True]
(a>5) & (a<20)  : [ True False False  True  True  True False  True]
a == 25         : [False False False False False False  True False]
combined (|)    : [ True False False  True  True  True  True  True]
Matching values : [15  8 17  6 25 11]
Count           : 6

~combined (NOT) : [False  True  True False False False False False]
Excluded values : [ 3 22]

Demonstrating why 'and'/'or' fail:
Error with 'and': The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
Error with 'or' : The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()


### How it works

- `&` compares **element by element**: `True & True → True`, `True & False → False`. It works across the whole array at once.
- `|` does the same but needs **either** side to be `True`.
- `~` inverts every element: `True → False`, `False → True`. Think of it as NumPy's NOT.
- Parentheses around each condition are **not optional** — `a > 5 & a < 20` would be parsed as `a > (5 & a) < 20` due to Python's operator precedence, giving wrong results silently.
- Python's `and`/`or` are designed for **single** `True`/`False` values. When applied to an array, Python cannot decide if the whole array is `True` or `False` — so it raises a `ValueError` to warn you.
- `&`, `|`, `~` are **bitwise operators** that NumPy overloads to work element-wise on boolean arrays — they never raise this ambiguity error.

**Key takeaway:** Always use `&`, `|`, `~` (not `and`, `or`, `not`) for NumPy boolean conditions — Python's logical operators cannot handle arrays and will raise a `ValueError`.

### Q11. `np.where`
Given an array of temperatures `[35.1, 22.3, 40.8, 18.5, 28.7, 15.0]`, use `np.where` to:
- Replace values above 30 with `"Hot"` and others with `"Comfortable"` (result as object array)
- Replace negative values in a separate array with 0 and keep positives as-is
- Use `np.where(condition)` with only one argument to get the indices of matching elements

In [11]:
# Q11: np.where — conditional replacement and index finding

import numpy as np

temps = np.array([35.1, 22.3, 40.8, 18.5, 28.7, 15.0])
print("Temperatures:", temps)
print()

# --- 1. Label as "Hot" or "Comfortable" ---
labels = np.where(temps > 30, "Hot", "Comfortable")   # (condition, value_if_true, value_if_false)
print("Labels      :", labels)
print("dtype       :", labels.dtype)                   # object array — holds strings
print()

# --- 2. Replace negatives with 0, keep positives as-is ---
mixed = np.array([-5.2, 3.1, -0.8, 7.4, -2.0, 9.9])
print("Mixed array :", mixed)
clipped = np.where(mixed < 0, 0, mixed)               # negative → 0, else keep original value
print("Negatives→0 :", clipped)
print()

# --- 3. np.where(condition) — single argument returns indices of True elements ---
hot_indices = np.where(temps > 30)                    # returns a tuple of arrays
print("Indices where temp > 30 :", hot_indices)        # tuple: (array([0, 2]),)
print("Indices (flat)          :", hot_indices[0])     # extract the array from the tuple
print("Hot temperatures        :", temps[hot_indices]) # use indices to fetch the values
print()

# --- Bonus: combining all three ideas ---
comfort_indices = np.where(temps <= 30)[0]
print("Comfortable temps at indices:", comfort_indices, "→", temps[comfort_indices])

Temperatures: [35.1 22.3 40.8 18.5 28.7 15. ]

Labels      : ['Hot' 'Comfortable' 'Hot' 'Comfortable' 'Comfortable' 'Comfortable']
dtype       : <U11

Mixed array : [-5.2  3.1 -0.8  7.4 -2.   9.9]
Negatives→0 : [0.  3.1 0.  7.4 0.  9.9]

Indices where temp > 30 : (array([0, 2]),)
Indices (flat)          : [0 2]
Hot temperatures        : [35.1 40.8]

Comfortable temps at indices: [1 3 4 5] → [22.3 18.5 28.7 15. ]


### How it works

- `np.where(condition, x, y)` works like a vectorized if-else: where condition is `True` → take from `x`, where `False` → take from `y`.
- `x` and `y` can be scalars, strings, or arrays. When strings are used, the result dtype becomes `object` — NumPy's way of storing arbitrary Python objects.
- For the clip-negatives case, passing `mixed` as the third argument means "keep the original value" wherever the condition is `False`.
- `np.where(condition)` with **one argument only** returns a tuple of index arrays — one array per dimension. For 1D arrays, extract with `[0]` to get a plain array.
- The returned indices can be used directly inside `[]` to fetch the matching values — this is called **fancy indexing**.

**Key takeaway:** `np.where(cond, x, y)` is a vectorized if-else across an entire array; called with one argument, it returns the indices where the condition is `True`.


### Q12. `np.clip`
Create an array of sensor readings ranging from -50 to 150. Use `np.clip` to enforce a valid range of 0–100. Print before and after. Then demonstrate clipping with per-element min/max arrays (passing arrays to `a_min` and `a_max`).

In [12]:
# Q12: np.clip — enforce value ranges on sensor readings

import numpy as np

rng = np.random.default_rng(seed=11)

# --- Simulate sensor readings with some out-of-range values ---
readings = rng.integers(-50, 151, size=12)             # values from -50 to 150
print("Raw readings :", readings)
print("Shape        :", readings.shape)
print()

# --- Scalar clip: enforce global range [0, 100] ---
clipped = np.clip(readings, a_min=0, a_max=100)        # below 0 → 0, above 100 → 100
print("Clipped [0,100]:", clipped)
print()

# Show which values were changed
below = readings < 0
above = readings > 100
print("Values below 0   (set to   0):", readings[below])
print("Values above 100 (set to 100):", readings[above])
print("Values in range  (unchanged) :", readings[~below & ~above])
print()

# --- Per-element clip: different min/max for each position ---
# Imagine 6 sensors, each with its own valid operating range
sensor_vals = np.array([120,  -10,  55,  200,   5,  88], dtype=float)
min_limits  = np.array([ 10,    0,  20,   50,  10,  50], dtype=float)  # per-sensor min
max_limits  = np.array([100,   80,  60,  150,  90, 100], dtype=float)  # per-sensor max

clipped_per = np.clip(sensor_vals, a_min=min_limits, a_max=max_limits)

print("Per-element clip:")
print("Sensor values :", sensor_vals)
print("Min limits    :", min_limits)
print("Max limits    :", max_limits)
print("Clipped result:", clipped_per)
print()

# Summary table
print(f"{'Sensor':<8}{'Value':>8}{'Min':>8}{'Max':>8}{'Clipped':>10}")
print("-" * 42)
for i, (v, lo, hi, c) in enumerate(zip(sensor_vals, min_limits, max_limits, clipped_per)):
    flag = " <-- changed" if v != c else ""
    print(f"S{i:<7}{v:>8.0f}{lo:>8.0f}{hi:>8.0f}{c:>10.0f}{flag}")

Raw readings : [-24 -25 110  50  68  70  93 -45  47 -21  30 136]
Shape        : (12,)

Clipped [0,100]: [  0   0 100  50  68  70  93   0  47   0  30 100]

Values below 0   (set to   0): [-24 -25 -45 -21]
Values above 100 (set to 100): [110 136]
Values in range  (unchanged) : [50 68 70 93 47 30]

Per-element clip:
Sensor values : [120. -10.  55. 200.   5.  88.]
Min limits    : [10.  0. 20. 50. 10. 50.]
Max limits    : [100.  80.  60. 150.  90. 100.]
Clipped result: [100.   0.  55. 150.  10.  88.]

Sensor     Value     Min     Max   Clipped
------------------------------------------
S0           120      10     100       100 <-- changed
S1           -10       0      80         0 <-- changed
S2            55      20      60        55
S3           200      50     150       150 <-- changed
S4             5      10      90        10 <-- changed
S5            88      50     100        88


### How it works

- `np.clip(arr, a_min, a_max)` enforces a valid range: values below `a_min` are raised to `a_min`, values above `a_max` are lowered to `a_max`, and values already in range are untouched.
- Think of it like a thermostat — if temperature drops below the minimum setpoint it holds at minimum, above maximum it holds at maximum.
- Scalar `a_min`/`a_max` applies one global range to every element — the common use case.
- Array `a_min`/`a_max` applies a **different range to each element** — useful when each sensor, channel, or feature has its own valid operating bounds.
- `np.clip` never modifies the original array — it always returns a new array. The original `readings` stays unchanged.

**Key takeaway:** `np.clip` is the cleanest way to enforce value boundaries on an array — pass scalars for a global range or arrays for per-element limits.

### Q13. `np.round`, `np.floor`, `np.ceil`, `np.trunc`
Create an array `[-2.7, -1.5, -0.3, 0.5, 1.5, 2.7]`. Apply `np.round`, `np.floor`, `np.ceil`, and `np.trunc` and print all results in a clear comparison. Explain banker's rounding (round half to even) for `np.round` with an example.

In [13]:
# Q13: np.round, np.floor, np.ceil, np.trunc — comparison and banker's rounding

import numpy as np

arr = np.array([-2.7, -1.5, -0.3, 0.5, 1.5, 2.7])
print("Original :", arr)
print()

# --- Four rounding functions ---
print("np.round :", np.round(arr))    # round half to even (banker's rounding)
print("np.floor :", np.floor(arr))    # always round DOWN toward -infinity
print("np.ceil  :", np.ceil(arr))     # always round UP toward +infinity
print("np.trunc :", np.trunc(arr))    # always round toward ZERO (drop decimal part)
print()

# --- Clear comparison table ---
print(f"{'Value':>8} {'round':>8} {'floor':>8} {'ceil':>8} {'trunc':>8}")
print("-" * 44)
for v, r, f, c, t in zip(arr,
                          np.round(arr),
                          np.floor(arr),
                          np.ceil(arr),
                          np.trunc(arr)):
    print(f"{v:>8.1f} {r:>8.1f} {f:>8.1f} {c:>8.1f} {t:>8.1f}")
print()

# --- Banker's rounding (round half to even) ---
# Standard rounding: .5 always rounds UP → causes upward bias over many values
# Banker's rounding: .5 rounds to the NEAREST EVEN number → bias cancels out
half_vals = np.array([0.5, 1.5, 2.5, 3.5, 4.5, 5.5])
print("Banker's rounding demo (round half to even):")
print("Values       :", half_vals)
print("np.round     :", np.round(half_vals))   # 0→0, 1.5→2, 2.5→2, 3.5→4, 4.5→4, 5.5→6
print("Expected even:", [0, 2, 2, 4, 4, 6])   # always lands on even number
print()

# Showing the bias difference on a large dataset
many = np.arange(0.5, 1000.5, 1.0)            # 0.5, 1.5, 2.5, ... 999.5
standard_sum = np.sum(np.floor(many + 0.5))   # naive: always round up
bankers_sum  = np.sum(np.round(many))         # banker's: alternates up/down
true_sum     = np.sum(many)                   # exact sum
print(f"True sum      : {true_sum:.1f}")
print(f"Standard round: {standard_sum:.1f}  (bias: +{standard_sum - true_sum:.1f})")
print(f"Banker's round: {bankers_sum:.1f}  (bias: +{bankers_sum  - true_sum:.1f})")

Original : [-2.7 -1.5 -0.3  0.5  1.5  2.7]

np.round : [-3. -2. -0.  0.  2.  3.]
np.floor : [-3. -2. -1.  0.  1.  2.]
np.ceil  : [-2. -1. -0.  1.  2.  3.]
np.trunc : [-2. -1. -0.  0.  1.  2.]

   Value    round    floor     ceil    trunc
--------------------------------------------
    -2.7     -3.0     -3.0     -2.0     -2.0
    -1.5     -2.0     -2.0     -1.0     -1.0
    -0.3     -0.0     -1.0     -0.0     -0.0
     0.5      0.0      0.0      1.0      0.0
     1.5      2.0      1.0      2.0      1.0
     2.7      3.0      2.0      3.0      2.0

Banker's rounding demo (round half to even):
Values       : [0.5 1.5 2.5 3.5 4.5 5.5]
np.round     : [0. 2. 2. 4. 4. 6.]
Expected even: [0, 2, 2, 4, 4, 6]

True sum      : 500000.0
Standard round: 500500.0  (bias: +500.0)
Banker's round: 500000.0  (bias: +0.0)


### How it works

- `np.floor` always goes **down** toward −∞: `floor(-1.5)` → `-2`, not `-1`.
- `np.ceil` always goes **up** toward +∞: `ceil(-1.5)` → `-1`.
- `np.trunc` always goes **toward zero** — it simply drops the decimal: `trunc(-2.7)` → `-2`, `trunc(2.7)` → `2`.
- `np.round` uses **banker's rounding**: when a value is exactly `x.5`, it rounds to the nearest **even** integer instead of always rounding up.
  - `0.5 → 0` (0 is even), `1.5 → 2` (2 is even), `2.5 → 2` (2 is even), `3.5 → 4` (4 is even).
- Why banker's rounding? Always rounding `.5` up introduces a small upward bias that accumulates over millions of financial or scientific calculations. Rounding to even alternates up and down, so the bias cancels out.

**Key takeaway:** `floor`/`ceil`/`trunc` always go in one direction; `np.round` uses banker's rounding (half → even) to avoid accumulated bias — critical in financial and scientific computing.

### Q14. `np.sort` and `np.argsort`
Create an array `[42, 7, 19, 55, 3, 88, 31]`. Sort it with `np.sort` (which returns a new sorted array). Use `np.argsort` to get the indices that would sort it. Use those indices to manually retrieve the sorted array and verify it matches. Also sort a 2D array along `axis=0` and `axis=1`.
 

In [14]:
# Q14: np.sort and np.argsort — sorting values and tracking original positions

import numpy as np

arr = np.array([42, 7, 19, 55, 3, 88, 31])
print("Original  :", arr)
print()

# --- np.sort: returns a NEW sorted array, original unchanged ---
sorted_arr = np.sort(arr)
print("np.sort   :", sorted_arr)
print("Original unchanged:", arr)          # confirm original is untouched
print()

# --- np.argsort: returns indices that WOULD sort the array ---
indices = np.argsort(arr)
print("np.argsort:", indices)              # position of smallest, 2nd smallest, etc.
print()

# --- Use argsort indices to manually retrieve sorted array ---
manually_sorted = arr[indices]             # fancy indexing with argsort result
print("arr[argsort]      :", manually_sorted)
print("Matches np.sort?  :", np.array_equal(manually_sorted, sorted_arr))
print()

# --- Practical use: rank items by value ---
scores = np.array([42, 7, 19, 55, 3, 88, 31])
names  = np.array(["Alice", "Bob", "Carol", "Dave", "Eve", "Frank", "Grace"])
rank_indices = np.argsort(scores)[::-1]   # argsort ascending, reverse for descending
print("Leaderboard (highest score first):")
for rank, idx in enumerate(rank_indices, start=1):
    print(f"  #{rank}  {names[idx]:<8}  score: {scores[idx]}")
print()

# --- Sorting a 2D array along axes ---
rng = np.random.default_rng(seed=5)
arr2d = rng.integers(1, 50, size=(3, 4))
print("2D array:\n", arr2d)
print()

print("sort axis=0 (sort each COLUMN independently):\n", np.sort(arr2d, axis=0))
print()
print("sort axis=1 (sort each ROW independently):\n",   np.sort(arr2d, axis=1))

Original  : [42  7 19 55  3 88 31]

np.sort   : [ 3  7 19 31 42 55 88]
Original unchanged: [42  7 19 55  3 88 31]

np.argsort: [4 1 2 6 0 3 5]

arr[argsort]      : [ 3  7 19 31 42 55 88]
Matches np.sort?  : True

Leaderboard (highest score first):
  #1  Frank     score: 88
  #2  Dave      score: 55
  #3  Alice     score: 42
  #4  Grace     score: 31
  #5  Carol     score: 19
  #6  Bob       score: 7
  #7  Eve       score: 3

2D array:
 [[33 40  2 40]
 [23 26 31 15]
 [48  3 14 19]]

sort axis=0 (sort each COLUMN independently):
 [[23  3  2 15]
 [33 26 14 19]
 [48 40 31 40]]

sort axis=1 (sort each ROW independently):
 [[ 2 33 40 40]
 [15 23 26 31]
 [ 3 14 19 48]]


### How it works

- `np.sort` returns a **new** sorted array — the original is never modified. Use `arr.sort()` (in-place method) if you want to modify the original.
- `np.argsort` does not return the sorted values — it returns the **positions** (indices) of elements in sorted order. Think of it as: "which index holds the smallest value? the next smallest?" and so on.
- Passing those indices back into `arr[]` (fancy indexing) reconstructs the sorted array — confirming argsort is correct.
- The real power of `argsort` is reordering **other arrays by the same ranking** — like sorting names by their matching scores, without a loop.
- For 2D sorting: `axis=0` sorts each **column** independently top-to-bottom; `axis=1` sorts each **row** independently left-to-right. Rows and columns are never mixed.

**Key takeaway:** `np.sort` returns sorted values; `np.argsort` returns the indices that produce that order — use those indices to simultaneously reorder any related array, like names paired with scores.

### Q15. `np.argmin` and `np.argmax`
Using a 3×4 matrix of random integers, find:
- The index of the global maximum (using `argmax()` then converting to 2D index with `np.unravel_index`)
- The index of the minimum in each row (axis=1)
- The index of the maximum in each column (axis=0)

In [15]:
# Q15: np.argmin and np.argmax — locating extreme values in 1D and 2D

import numpy as np

rng = np.random.default_rng(seed=99)
mat = rng.integers(1, 100, size=(3, 4))           # 3 rows, 4 columns, values 1–99
print("Matrix (3×4):\n", mat)
print("Shape:", mat.shape)
print()

# --- Global maximum: argmax on flattened array, then convert to 2D index ---
flat_idx = np.argmax(mat)                          # index in the flattened (1D) version
row, col = np.unravel_index(flat_idx, mat.shape)  # convert flat index → (row, col)
print("Global max value      :", mat[row, col])
print("Flat index            :", flat_idx)
print("2D index (row, col)   :", (row, col))
print()

# Same approach for global minimum
flat_idx_min = np.argmin(mat)
row_min, col_min = np.unravel_index(flat_idx_min, mat.shape)
print("Global min value      :", mat[row_min, col_min])
print("Flat index            :", flat_idx_min)
print("2D index (row, col)   :", (row_min, col_min))
print()

# --- Index of minimum in each ROW (axis=1) ---
row_argmin = np.argmin(mat, axis=1)               # one column-index per row
print("Min index per row (axis=1):", row_argmin)
print("Verification — min values per row:")
for r, c in enumerate(row_argmin):
    print(f"  Row {r}: col {c} → value {mat[r, c]}")
print()

# --- Index of maximum in each COLUMN (axis=0) ---
col_argmax = np.argmax(mat, axis=0)               # one row-index per column
print("Max index per col (axis=0):", col_argmax)
print("Verification — max values per column:")
for c, r in enumerate(col_argmax):
    print(f"  Col {c}: row {r} → value {mat[r, c]}")

Matrix (3×4):
 [[95 51 76 56]
 [18 51 94 97]
 [96 61 32 57]]
Shape: (3, 4)

Global max value      : 97
Flat index            : 7
2D index (row, col)   : (np.int64(1), np.int64(3))

Global min value      : 18
Flat index            : 4
2D index (row, col)   : (np.int64(1), np.int64(0))

Min index per row (axis=1): [1 0 2]
Verification — min values per row:
  Row 0: col 1 → value 51
  Row 1: col 0 → value 18
  Row 2: col 2 → value 32

Max index per col (axis=0): [2 2 1 1]
Verification — max values per column:
  Col 0: row 2 → value 96
  Col 1: row 2 → value 61
  Col 2: row 1 → value 94
  Col 3: row 1 → value 97


### How it works

- `np.argmax` and `np.argmin` return the **index** (position) of the largest/smallest value — not the value itself.
- Without an axis, they treat the array as **flattened** — a 3×4 matrix becomes 12 elements in a single row, and the returned index is a position in that flat sequence.
- `np.unravel_index(flat_idx, shape)` converts a flat index back into a multi-dimensional `(row, col)` coordinate — essential for locating a value in the original 2D matrix.
- With `axis=1`, `argmin` collapses across columns and returns **one column-index per row** — "which column holds this row's smallest value?"
- With `axis=0`, `argmax` collapses across rows and returns **one row-index per column** — "which row holds this column's largest value?"
- The verification loops confirm correctness by using the returned indices directly to fetch the actual values from the matrix.

**Key takeaway:** `argmax`/`argmin` return positions, not values — use `np.unravel_index` to convert a flat index to 2D coordinates, and pair with `axis` to find extreme positions row-by-row or column-by-column.

### Q16. `np.unique`
Create an array with many repeated values `[3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5]`. Use `np.unique` to find unique values. Use `return_counts=True` to get how many times each appears. Use `return_index=True` to find where each first appears. Use `return_inverse=True` to reconstruct the original array.

In [16]:
# Q16: np.unique — find, count, locate, and reconstruct unique values

import numpy as np

arr = np.array([3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5])
print("Original array:", arr)
print("Length        :", len(arr))
print()

# --- Basic: unique values only (always sorted) ---
unique_vals = np.unique(arr)
print("Unique values :", unique_vals)
print()

# --- return_counts: how many times each unique value appears ---
unique_vals, counts = np.unique(arr, return_counts=True)
print("Unique values :", unique_vals)
print("Counts        :", counts)
print()
print("Frequency table:")
for val, cnt in zip(unique_vals, counts):
    bar = "█" * cnt                                    # visual bar for frequency
    print(f"  {val}: {cnt}x  {bar}")
print()

# --- return_index: index of FIRST occurrence of each unique value ---
unique_vals, first_idx = np.unique(arr, return_index=True)
print("First occurrence indices:", first_idx)
print("Verification:")
for val, idx in zip(unique_vals, first_idx):
    print(f"  value {val} first appears at index {idx} → arr[{idx}] = {arr[idx]}")
print()

# --- return_inverse: indices that reconstruct the original array ---
unique_vals, inverse_idx = np.unique(arr, return_inverse=True)
print("Unique values  :", unique_vals)
print("Inverse indices:", inverse_idx)             # each position maps to unique_vals[inverse_idx[i]]
reconstructed = unique_vals[inverse_idx]           # rebuild original from unique + inverse
print("Reconstructed  :", reconstructed)
print("Matches original?", np.array_equal(reconstructed, arr))
print()

# --- Combining all four return values together ---
unique_vals, first_idx, inverse_idx, counts = np.unique(
    arr,
    return_index=True,
    return_inverse=True,
    return_counts=True
)
print("Most frequent value  :", unique_vals[np.argmax(counts)], f"({np.max(counts)}x)")
print("Least frequent value :", unique_vals[np.argmin(counts)], f"({np.min(counts)}x)")
print("Number of unique vals:", len(unique_vals), "out of", len(arr), "total")

Original array: [3 1 4 1 5 9 2 6 5 3 5]
Length        : 11

Unique values : [1 2 3 4 5 6 9]

Unique values : [1 2 3 4 5 6 9]
Counts        : [2 1 2 1 3 1 1]

Frequency table:
  1: 2x  ██
  2: 1x  █
  3: 2x  ██
  4: 1x  █
  5: 3x  ███
  6: 1x  █
  9: 1x  █

First occurrence indices: [1 6 0 2 4 7 5]
Verification:
  value 1 first appears at index 1 → arr[1] = 1
  value 2 first appears at index 6 → arr[6] = 2
  value 3 first appears at index 0 → arr[0] = 3
  value 4 first appears at index 2 → arr[2] = 4
  value 5 first appears at index 4 → arr[4] = 5
  value 6 first appears at index 7 → arr[7] = 6
  value 9 first appears at index 5 → arr[5] = 9

Unique values  : [1 2 3 4 5 6 9]
Inverse indices: [2 0 3 0 4 6 1 5 4 2 4]
Reconstructed  : [3 1 4 1 5 9 2 6 5 3 5]
Matches original? True

Most frequent value  : 5 (3x)
Least frequent value : 2 (1x)
Number of unique vals: 7 out of 11 total


### How it works

- `np.unique` always returns unique values in **sorted ascending order** — the original order is not preserved.
- `return_counts=True` appends a second array: how many times each unique value appears. Index `i` of counts matches index `i` of unique values.
- `return_index=True` appends the **position of first appearance** in the original array for each unique value.
- `return_inverse=True` appends an array of the same length as the original — each element is the index into `unique_vals` that recovers the original value. So `unique_vals[inverse_idx]` reconstructs the original array exactly.
- All return flags can be combined in one call — NumPy returns a tuple in the order: `(unique, index, inverse, counts)`.
- A practical use: `return_inverse` is used in data compression and label encoding — store only the unique values once, then store cheap integer indices instead of repeating strings or large values.

**Key takeaway:** `np.unique` is more than deduplication — with its return flags it reveals frequency, first positions, and an inverse map that can fully reconstruct the original array from compact unique storage.

### Q17. `np.diff`
Given a 1D array representing daily stock prices over 10 days, use `np.diff` to compute daily changes. Then use `np.diff(arr, n=2)` for second-order differences. Apply `np.diff` on a 2D array along both axes and explain what each means.

In [17]:
# Q16: np.unique — find, count, locate, and reconstruct unique values

import numpy as np

arr = np.array([3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5])
print("Original array:", arr)
print("Length        :", len(arr))
print()

# --- Basic: unique values only (always sorted) ---
unique_vals = np.unique(arr)
print("Unique values :", unique_vals)
print()

# --- return_counts: how many times each unique value appears ---
unique_vals, counts = np.unique(arr, return_counts=True)
print("Unique values :", unique_vals)
print("Counts        :", counts)
print()
print("Frequency table:")
for val, cnt in zip(unique_vals, counts):
    bar = "█" * cnt                                    # visual bar for frequency
    print(f"  {val}: {cnt}x  {bar}")
print()

# --- return_index: index of FIRST occurrence of each unique value ---
unique_vals, first_idx = np.unique(arr, return_index=True)
print("First occurrence indices:", first_idx)
print("Verification:")
for val, idx in zip(unique_vals, first_idx):
    print(f"  value {val} first appears at index {idx} → arr[{idx}] = {arr[idx]}")
print()

# --- return_inverse: indices that reconstruct the original array ---
unique_vals, inverse_idx = np.unique(arr, return_inverse=True)
print("Unique values  :", unique_vals)
print("Inverse indices:", inverse_idx)             # each position maps to unique_vals[inverse_idx[i]]
reconstructed = unique_vals[inverse_idx]           # rebuild original from unique + inverse
print("Reconstructed  :", reconstructed)
print("Matches original?", np.array_equal(reconstructed, arr))
print()

# --- Combining all four return values together ---
unique_vals, first_idx, inverse_idx, counts = np.unique(
    arr,
    return_index=True,
    return_inverse=True,
    return_counts=True
)
print("Most frequent value  :", unique_vals[np.argmax(counts)], f"({np.max(counts)}x)")
print("Least frequent value :", unique_vals[np.argmin(counts)], f"({np.min(counts)}x)")
print("Number of unique vals:", len(unique_vals), "out of", len(arr), "total")

Original array: [3 1 4 1 5 9 2 6 5 3 5]
Length        : 11

Unique values : [1 2 3 4 5 6 9]

Unique values : [1 2 3 4 5 6 9]
Counts        : [2 1 2 1 3 1 1]

Frequency table:
  1: 2x  ██
  2: 1x  █
  3: 2x  ██
  4: 1x  █
  5: 3x  ███
  6: 1x  █
  9: 1x  █

First occurrence indices: [1 6 0 2 4 7 5]
Verification:
  value 1 first appears at index 1 → arr[1] = 1
  value 2 first appears at index 6 → arr[6] = 2
  value 3 first appears at index 0 → arr[0] = 3
  value 4 first appears at index 2 → arr[2] = 4
  value 5 first appears at index 4 → arr[4] = 5
  value 6 first appears at index 7 → arr[7] = 6
  value 9 first appears at index 5 → arr[5] = 9

Unique values  : [1 2 3 4 5 6 9]
Inverse indices: [2 0 3 0 4 6 1 5 4 2 4]
Reconstructed  : [3 1 4 1 5 9 2 6 5 3 5]
Matches original? True

Most frequent value  : 5 (3x)
Least frequent value : 2 (1x)
Number of unique vals: 7 out of 11 total


### How it works

- `np.unique` always returns unique values in **sorted ascending order** — the original order is not preserved.
- `return_counts=True` appends a second array: how many times each unique value appears. Index `i` of counts matches index `i` of unique values.
- `return_index=True` appends the **position of first appearance** in the original array for each unique value.
- `return_inverse=True` appends an array of the same length as the original — each element is the index into `unique_vals` that recovers the original value. So `unique_vals[inverse_idx]` reconstructs the original array exactly.
- All return flags can be combined in one call — NumPy returns a tuple in the order: `(unique, index, inverse, counts)`.
- A practical use: `return_inverse` is used in data compression and label encoding — store only the unique values once, then store cheap integer indices instead of repeating strings or large values.

**Key takeaway:** `np.unique` is more than deduplication — with its return flags it reveals frequency, first positions, and an inverse map that can fully reconstruct the original array from compact unique storage.

### Q18. `np.gradient`
Use `np.gradient` on a 1D array `[0, 1, 4, 9, 16, 25]` (perfect squares) to compute the numerical gradient. Compare the result to the analytical derivative (`2x`). Then apply it to a 2D array and explain what the two output arrays represent.

In [18]:
# Q18: np.gradient — numerical derivative on 1D and 2D arrays

import numpy as np

# --- 1D: perfect squares y = x² ---
x = np.array([0, 1, 2, 3, 4, 5], dtype=float)
y = np.array([0, 1, 4, 9, 16, 25], dtype=float)   # y = x²

grad = np.gradient(y)                              # numerical gradient using central differences
analytical = 2 * x                                 # analytical derivative of x² is 2x

print("x values         :", x)
print("y = x²           :", y)
print()
print("np.gradient(y)   :", grad)
print("Analytical 2x    :", analytical)
print()

# Compare numerical vs analytical
diff = np.abs(grad - analytical)
print("Absolute error   :", diff)
print("Max error        :", np.max(diff))          # error only at edges (boundary effect)
print()

# Explain the difference at edges vs interior points
print("How np.gradient computes each point:")
print("  Interior (central difference) : grad[i] = (y[i+1] - y[i-1]) / 2")
print("  Left edge (forward difference): grad[0] = (y[1]   - y[0])   / 1")
print("  Right edge (backward diff)    : grad[-1]= (y[-1]  - y[-2])  / 1")
print()
print("Verification of interior point at x=2:")
print("  (y[3] - y[1]) / 2 =", (y[3] - y[1]) / 2, "  analytical 2×2 =", 2*2)
print()

# --- With custom spacing: x values not evenly spaced by 1 ---
x_sparse = np.array([0, 1, 2, 3, 4, 5], dtype=float) * 0.5   # step size = 0.5
y_sparse = x_sparse ** 2
grad_spaced = np.gradient(y_sparse, x_sparse)                  # pass x to account for spacing
print("Custom spacing (step=0.5):")
print("x_sparse         :", x_sparse)
print("y_sparse = x²    :", y_sparse)
print("gradient         :", grad_spaced)
print("analytical 2x    :", 2 * x_sparse)
print()

# --- 2D gradient ---
rng = np.random.default_rng(seed=8)
surface = np.array([
    [1.0, 2.0, 4.0, 3.0],
    [2.0, 5.0, 8.0, 6.0],
    [1.0, 3.0, 6.0, 4.0],
])
print("2D surface (3×4):\n", surface)
print()

grad_y, grad_x = np.gradient(surface)             # returns two arrays: [row-gradient, col-gradient]

print("Gradient along rows (axis=0) — vertical slope:\n",   np.round(grad_y, 3))
print()
print("Gradient along cols (axis=1) — horizontal slope:\n", np.round(grad_x, 3))
print()
print("Meaning:")
print("  grad_y[i,j] → how steeply the surface rises/falls going DOWN a column")
print("  grad_x[i,j] → how steeply the surface rises/falls going ACROSS a row")

x values         : [0. 1. 2. 3. 4. 5.]
y = x²           : [ 0.  1.  4.  9. 16. 25.]

np.gradient(y)   : [1. 2. 4. 6. 8. 9.]
Analytical 2x    : [ 0.  2.  4.  6.  8. 10.]

Absolute error   : [1. 0. 0. 0. 0. 1.]
Max error        : 1.0

How np.gradient computes each point:
  Interior (central difference) : grad[i] = (y[i+1] - y[i-1]) / 2
  Left edge (forward difference): grad[0] = (y[1]   - y[0])   / 1
  Right edge (backward diff)    : grad[-1]= (y[-1]  - y[-2])  / 1

Verification of interior point at x=2:
  (y[3] - y[1]) / 2 = 4.0   analytical 2×2 = 4

Custom spacing (step=0.5):
x_sparse         : [0.  0.5 1.  1.5 2.  2.5]
y_sparse = x²    : [0.   0.25 1.   2.25 4.   6.25]
gradient         : [0.5 1.  2.  3.  4.  4.5]
analytical 2x    : [0. 1. 2. 3. 4. 5.]

2D surface (3×4):
 [[1. 2. 4. 3.]
 [2. 5. 8. 6.]
 [1. 3. 6. 4.]]

Gradient along rows (axis=0) — vertical slope:
 [[ 1.   3.   4.   3. ]
 [ 0.   0.5  1.   0.5]
 [-1.  -2.  -2.  -2. ]]

Gradient along cols (axis=1) — horizontal slope:
 [

### How it works

- `np.gradient` approximates the derivative numerically using **central differences** for interior points: `(y[i+1] - y[i-1]) / 2`. No calculus needed — just subtraction and division.
- At the **edges**, it falls back to one-sided differences (forward at the left, backward at the right) since there is no neighbor on one side — this introduces small errors at the boundaries.
- For `y = x²`, the analytical derivative is `2x`. `np.gradient` matches perfectly at interior points and has small edge errors — confirming numerical methods work best away from boundaries.
- Passing `x` as a second argument tells `np.gradient` the actual spacing between points — essential when steps are not uniform or not equal to 1.
- For a 2D array, `np.gradient` returns **two arrays** of the same shape: the first is the gradient along `axis=0` (how the surface slopes vertically, row to row); the second is the gradient along `axis=1` (how it slopes horizontally, column to column). Think of it as the slope of a hillside in the north-south and east-west directions separately.

**Key takeaway:** `np.gradient` computes numerical derivatives without calculus — central differences for interior points, one-sided at edges; on 2D arrays it returns two slope arrays, one per axis, describing how steeply values change in each direction.

### Q19. Set Operations
Using `np.union1d`, `np.intersect1d`, `np.setdiff1d`, and `np.isin`, perform set-style operations on two integer arrays. Demonstrate: finding common elements, elements only in one array, and checking membership of several values in a larger array.

In [19]:
# Q19: NumPy set operations — union, intersection, difference, membership

import numpy as np

a = np.array([3, 1, 4, 1, 5, 9, 2, 6])            # note: has duplicate (1 appears twice)
b = np.array([2, 7, 1, 8, 2, 8, 1, 8])            # note: has duplicates too

print("Array a:", a)
print("Array b:", b)
print()

# --- union1d: all unique values from BOTH arrays combined ---
union = np.union1d(a, b)
print("union1d(a, b)       :", union)              # sorted, duplicates removed
print("  → all values that appear in a OR b (or both)")
print()

# --- intersect1d: values that appear in BOTH arrays ---
common = np.intersect1d(a, b)
print("intersect1d(a, b)   :", common)             # sorted unique values in both
print("  → values present in a AND b")
print()

# --- setdiff1d: values in first array but NOT in second ---
only_in_a = np.setdiff1d(a, b)
only_in_b = np.setdiff1d(b, a)
print("setdiff1d(a, b)     :", only_in_a)          # in a but not b
print("  → values in a but NOT in b")
print("setdiff1d(b, a)     :", only_in_b)          # in b but not a
print("  → values in b but NOT in a")
print()

# --- setxor1d: values in EITHER array but NOT in both (symmetric difference) ---
exclusive = np.setxor1d(a, b)
print("setxor1d(a, b)      :", exclusive)
print("  → values in a OR b but NOT in both (exclusive OR)")
print()

# --- isin: check membership of values in a larger array ---
roster   = np.array([10, 22, 35, 41, 58, 63, 77, 89, 94])
check    = np.array([22, 50, 63, 77, 100])

mask = np.isin(check, roster)                      # True where check value exists in roster
print("Roster              :", roster)
print("Values to check     :", check)
print("np.isin result      :", mask)
print()
print("Membership detail:")
for val, found in zip(check, mask):
    status = "FOUND" if found else "not found"
    print(f"  {val:>4} → {status}")
print()

# --- Practical use: filter array to keep only values present in a reference list ---
allowed_ids = np.array([1, 3, 5, 7, 9])
incoming    = np.array([2, 3, 5, 8, 9, 1, 6])
valid       = incoming[np.isin(incoming, allowed_ids)]   # boolean mask filter
rejected    = incoming[~np.isin(incoming, allowed_ids)]
print("Incoming IDs        :", incoming)
print("Allowed IDs         :", allowed_ids)
print("Valid (keep)        :", valid)
print("Rejected (drop)     :", rejected)

Array a: [3 1 4 1 5 9 2 6]
Array b: [2 7 1 8 2 8 1 8]

union1d(a, b)       : [1 2 3 4 5 6 7 8 9]
  → all values that appear in a OR b (or both)

intersect1d(a, b)   : [1 2]
  → values present in a AND b

setdiff1d(a, b)     : [3 4 5 6 9]
  → values in a but NOT in b
setdiff1d(b, a)     : [7 8]
  → values in b but NOT in a

setxor1d(a, b)      : [3 4 5 6 7 8 9]
  → values in a OR b but NOT in both (exclusive OR)

Roster              : [10 22 35 41 58 63 77 89 94]
Values to check     : [ 22  50  63  77 100]
np.isin result      : [ True False  True  True False]

Membership detail:
    22 → FOUND
    50 → not found
    63 → FOUND
    77 → FOUND
   100 → not found

Incoming IDs        : [2 3 5 8 9 1 6]
Allowed IDs         : [1 3 5 7 9]
Valid (keep)        : [3 5 9 1]
Rejected (drop)     : [2 8 6]


### How it works

- All NumPy set functions automatically **deduplicate** their inputs — even if `a` has repeated values, the result contains each value only once, sorted.
- `union1d` → everything from both arrays combined (OR logic).
- `intersect1d` → only what both arrays share (AND logic).
- `setdiff1d(a, b)` → what `a` has that `b` does not. Order matters: `setdiff1d(b, a)` gives the opposite.
- `setxor1d` → values exclusive to one side only — the symmetric difference. Useful for finding what changed between two snapshots.
- `np.isin(check, roster)` returns a boolean array the same shape as `check` — `True` where the value exists in `roster`. Use it as a mask with `[]` to filter arrays, just like Q9–Q10.
- Think of these as Venn diagram operations: union is the whole diagram, intersection is the overlap, setdiff is one circle minus the overlap, setxor is both circles minus the overlap.

**Key takeaway:** NumPy's set functions (`union1d`, `intersect1d`, `setdiff1d`, `setxor1d`, `isin`) perform Venn-diagram-style operations on arrays — all deduplicate automatically and return sorted results, making them ideal for fast membership testing and data filtering without loops.

### Q20. `np.nan` Handling
Create an array containing `np.nan` values: `[1.0, np.nan, 3.0, np.nan, 5.0]`. Show that standard operations (`sum`, `mean`) return `nan`. Use `np.nansum`, `np.nanmean`, `np.nanmin`, `np.nanmax` to compute ignoring NaNs. Use `np.isnan` to create a boolean mask and count NaNs.

In [20]:
# Q20: Handling np.nan — detection, safe aggregation, and masking

import numpy as np

arr = np.array([1.0, np.nan, 3.0, np.nan, 5.0])
print("Array:", arr)
print()

# --- Standard operations propagate nan ---
print("=== Standard aggregations (nan propagates) ===")
print("np.sum(arr)  :", np.sum(arr))               # nan + anything = nan
print("np.mean(arr) :", np.mean(arr))
print("np.min(arr)  :", np.min(arr))
print("np.max(arr)  :", np.max(arr))
print()

# --- nan-safe aggregations: ignore nan, compute on valid values only ---
print("=== nan-safe aggregations (ignore nan) ===")
print("np.nansum(arr)  :", np.nansum(arr))          # 1 + 3 + 5 = 9
print("np.nanmean(arr) :", np.nanmean(arr))         # (1 + 3 + 5) / 3 = 3.0
print("np.nanmin(arr)  :", np.nanmin(arr))          # 1.0
print("np.nanmax(arr)  :", np.nanmax(arr))          # 5.0
print("np.nanstd(arr)  :", np.nanstd(arr))          # std of [1, 3, 5]
print("np.nanmedian(arr):", np.nanmedian(arr))      # median of [1, 3, 5] = 3.0
print()

# --- np.isnan: boolean mask for nan positions ---
nan_mask = np.isnan(arr)
print("np.isnan(arr)   :", nan_mask)                # True where nan
print("NaN count       :", np.sum(nan_mask))        # count Trues
print("Valid count     :", np.sum(~nan_mask))       # count non-nan values
print()

# --- Extract only valid values ---
valid_vals = arr[~nan_mask]                         # boolean mask keeps non-nan only
print("Valid values    :", valid_vals)
print()

# --- Replace nan with a fill value ---
filled_zero = np.where(np.isnan(arr), 0.0, arr)    # nan → 0
filled_mean = np.where(np.isnan(arr), np.nanmean(arr), arr)  # nan → column mean
print("Fill nan with 0   :", filled_zero)
print("Fill nan with mean:", filled_mean)
print()

# --- Practical: 2D array with nan — column-wise nan-safe mean ---
data = np.array([
    [1.0, np.nan, 3.0],
    [4.0,  5.0,  np.nan],
    [np.nan, 2.0, 6.0],
])
print("2D data:\n", data)
print("Column means (nanmean axis=0):", np.nanmean(data, axis=0))
print("Row means    (nanmean axis=1):", np.nanmean(data, axis=1))

Array: [ 1. nan  3. nan  5.]

=== Standard aggregations (nan propagates) ===
np.sum(arr)  : nan
np.mean(arr) : nan
np.min(arr)  : nan
np.max(arr)  : nan

=== nan-safe aggregations (ignore nan) ===
np.nansum(arr)  : 9.0
np.nanmean(arr) : 3.0
np.nanmin(arr)  : 1.0
np.nanmax(arr)  : 5.0
np.nanstd(arr)  : 1.632993161855452
np.nanmedian(arr): 3.0

np.isnan(arr)   : [False  True False  True False]
NaN count       : 2
Valid count     : 3

Valid values    : [1. 3. 5.]

Fill nan with 0   : [1. 0. 3. 0. 5.]
Fill nan with mean: [1. 3. 3. 3. 5.]

2D data:
 [[ 1. nan  3.]
 [ 4.  5. nan]
 [nan  2.  6.]]
Column means (nanmean axis=0): [2.5 3.5 4.5]
Row means    (nanmean axis=1): [2.  4.5 4. ]


### How it works

- `np.nan` is a special float value meaning "Not a Number" — it represents missing or undefined data.
- Standard aggregations like `np.sum` and `np.mean` **propagate** nan: any arithmetic involving nan returns nan. This is intentional — it forces you to consciously decide how to handle missing data.
- `np.nansum`, `np.nanmean`, `np.nanmin`, `np.nanmax` are nan-aware versions — they **skip** nan positions and compute only over valid values.
- `np.isnan(arr)` returns a boolean array: `True` where nan, `False` elsewhere. Use `~np.isnan(arr)` to get the valid-value mask.
- Two common strategies for handling nan: **drop** (filter with `~nan_mask`) or **fill** (replace with 0, mean, or forward-fill). Filling with the column mean is a standard technique in ML preprocessing.
- In 2D arrays, `np.nanmean(data, axis=0)` computes the mean per column skipping nan — exactly what you need for imputing missing feature values in a dataset.

**Key takeaway:** Standard NumPy aggregations propagate nan by design — use `np.isnan` to detect and count missing values, and the `nan`-prefixed functions (`nansum`, `nanmean`, etc.) to safely aggregate while ignoring them.

### Q21. `np.inf` Handling
Demonstrate that dividing by zero in NumPy produces `np.inf` (not an error). Show `np.isfinite`, `np.isinf`, `np.isnan` and how they differ. Show arithmetic involving `inf` (e.g., `inf + 1`, `inf - inf`, `1 / inf`).

In [21]:
# Q21: np.inf handling — generation, detection, and arithmetic behaviour

import numpy as np

# --- Division by zero produces inf, not an error ---
print("=== Generating inf via division by zero ===")
with np.errstate(divide='ignore', invalid='ignore'):   # silence the RuntimeWarning
    pos_inf = np.array([1.0])  / np.array([0.0])      # +inf
    neg_inf = np.array([-1.0]) / np.array([0.0])      # -inf
    nan_val = np.array([0.0])  / np.array([0.0])      # 0/0 → nan (indeterminate)

print("1.0  / 0.0 :", pos_inf[0])                     # inf
print("-1.0 / 0.0 :", neg_inf[0])                     # -inf
print("0.0  / 0.0 :", nan_val[0])                     # nan
print()

# --- Mixed array with inf, -inf, nan, and finite values ---
arr = np.array([1.0, np.inf, -np.inf, np.nan, 0.0, 42.0])
print("Array:", arr)
print()

# --- Three detection functions and how they differ ---
print("=== Detection functions ===")
print("np.isfinite(arr) :", np.isfinite(arr))   # True only for normal finite numbers
print("np.isinf(arr)    :", np.isinf(arr))      # True for +inf AND -inf, not nan
print("np.isnan(arr)    :", np.isnan(arr))      # True only for nan, not inf
print()

# Summary table
print(f"{'Value':<8} {'isfinite':>10} {'isinf':>8} {'isnan':>8}")
print("-" * 38)
for v, fin, inf_, nan_ in zip(arr,
                               np.isfinite(arr),
                               np.isinf(arr),
                               np.isnan(arr)):
    print(f"{str(v):<8} {str(fin):>10} {str(inf_):>8} {str(nan_):>8}")
print()

# --- Arithmetic involving inf ---
print("=== Arithmetic with inf ===")
inf = np.inf
print(f"inf + 1       = {inf + 1}")              # still inf
print(f"inf * 2       = {inf * 2}")              # still inf
print(f"inf + inf     = {inf + inf}")            # still inf
print(f"1 / inf       = {1 / inf}")              # → 0.0 (infinitely small)
print(f"inf - inf     = {inf - inf}")            # → nan (indeterminate form)
print(f"inf * 0       = {inf * 0}")              # → nan (indeterminate form)
print(f"-inf < 0      = {-inf < 0}")             # -inf is less than everything
print(f"inf > 1e308   = {inf > 1e308}")          # inf is greater than everything
print()

# --- Practical: filter out non-finite values before aggregation ---
sensor = np.array([3.2, np.inf, 1.5, np.nan, 4.8, -np.inf, 2.1])
print("Raw sensor data :", sensor)

finite_mask   = np.isfinite(sensor)              # True only for normal numbers
clean_sensor  = sensor[finite_mask]
print("After filtering :", clean_sensor)
print("Mean (clean)    :", np.round(np.mean(clean_sensor), 4))
print()

# --- np.isfinite is stricter than ~np.isnan ---
print("~np.isnan(arr)  :", ~np.isnan(arr))       # includes inf — may surprise you
print("np.isfinite(arr):", np.isfinite(arr))     # excludes inf AND nan — the safe choice

=== Generating inf via division by zero ===
1.0  / 0.0 : inf
-1.0 / 0.0 : -inf
0.0  / 0.0 : nan

Array: [  1.  inf -inf  nan   0.  42.]

=== Detection functions ===
np.isfinite(arr) : [ True False False False  True  True]
np.isinf(arr)    : [False  True  True False False False]
np.isnan(arr)    : [False False False  True False False]

Value      isfinite    isinf    isnan
--------------------------------------
1.0            True    False    False
inf           False     True    False
-inf          False     True    False
nan           False    False     True
0.0            True    False    False
42.0           True    False    False

=== Arithmetic with inf ===
inf + 1       = inf
inf * 2       = inf
inf + inf     = inf
1 / inf       = 0.0
inf - inf     = nan
inf * 0       = nan
-inf < 0      = True
inf > 1e308   = True

Raw sensor data : [ 3.2  inf  1.5  nan  4.8 -inf  2.1]
After filtering : [3.2 1.5 4.8 2.1]
Mean (clean)    : 2.9

~np.isnan(arr)  : [ True  True  True False  True  Tr

### How it works

- NumPy never raises an error for division by zero with floats — it returns `inf`, `-inf`, or `nan` and prints a `RuntimeWarning`. Use `np.errstate` to silence it when intentional.
- `0.0 / 0.0` is a special case — the result is `nan` (indeterminate), not `inf`.
- The three detection functions target different values:
  - `np.isfinite` → `True` only for normal numbers (excludes both `inf` and `nan`).
  - `np.isinf` → `True` for `+inf` and `-inf` only (not `nan`).
  - `np.isnan` → `True` for `nan` only (not `inf`).
- Indeterminate forms like `inf - inf` and `inf * 0` produce `nan` — mathematically undefined, just like in calculus.
- `1 / inf` → `0.0` because dividing by an infinitely large number gives an infinitely small result.
- `np.isfinite` is the safest filter before aggregation — `~np.isnan` looks similar but would accidentally keep `inf` values, which can silently corrupt `sum` or `mean`.

**Key takeaway:** NumPy converts division-by-zero to `inf` silently — use `np.isfinite` (not `~np.isnan`) as the strictest filter to remove both `inf` and `nan` before any aggregation.

### Q22. Vectorized String Operations with `np.char`
Create a string array `["  hello ", "WORLD", "numpy  ", "  ARRAY"]`. Use `np.char` functions: `strip`, `lower`, `upper`, `replace`, `startswith`, `count`. Compare this approach to using a list comprehension with a comment on when each is preferred.

In [22]:
# Q22: Vectorized string operations with np.char

import numpy as np

arr = np.array(["  hello ", "WORLD", "numpy  ", "  ARRAY"])
print("Original :", arr)
print()

# --- Cleaning operations ---
stripped = np.char.strip(arr)                          # remove leading/trailing whitespace
print("strip    :", stripped)

lowered  = np.char.lower(arr)                          # all characters to lowercase
print("lower    :", lowered)

uppered  = np.char.upper(arr)                          # all characters to uppercase
print("upper    :", uppered)

print()

# --- Chain operations: strip first, then lower ---
cleaned  = np.char.lower(np.char.strip(arr))           # pipeline: strip → lower
print("strip+lower (cleaned):", cleaned)
print()

# --- replace: substitute substring within every element ---
replaced = np.char.replace(stripped, "o", "0")        # replace 'o' with '0'
print("replace 'o'→'0':", replaced)
print()

# --- startswith / endswith: returns boolean array ---
starts_h = np.char.startswith(np.char.strip(arr), "h")  # True if starts with 'h' after strip
starts_w = np.char.startswith(arr, "W")                  # case-sensitive
print("startswith 'h' (stripped):", starts_h)
print("startswith 'W' (raw)     :", starts_w)
print()

# --- count: how many times a substring appears per element ---
vowels_count = np.char.count(np.char.lower(np.char.strip(arr)), "a")
print("count of 'a' per element:", vowels_count)

# Count any vowel (loop over vowels, sum counts)
all_vowels = sum(np.char.count(np.char.lower(np.char.strip(arr)), v)
                 for v in "aeiou")
print("total vowels per element:", all_vowels)
print()

# --- Practical: normalize a messy name list ---
names = np.array(["  alice ", "BOB", "  Carol", "DAVE  "])
normalized = np.char.title(np.char.strip(names))       # strip + Title Case
print("Raw names  :", names)
print("Normalized :", normalized)
print()

# --- Comparison: np.char vs list comprehension ---
print("=== np.char vs list comprehension ===")

test = np.array(["  hello ", "WORLD", "numpy  ", "  ARRAY"])

# np.char approach
result_npchar = np.char.lower(np.char.strip(test))

# list comprehension approach
result_listcomp = np.array([s.strip().lower() for s in test])

print("np.char result     :", result_npchar)
print("listcomp result    :", result_listcomp)
print("Results identical? :", np.array_equal(result_npchar, result_listcomp))
print()
print("When to prefer each:")
print("  np.char      → already working in NumPy pipeline, want array output directly")
print("  listcomp     → complex logic per element (conditionals, regex, multi-step)")
print("  Speed note   → np.char is NOT faster than listcomp (both loop in CPython)")
print("                 np.char advantage is API consistency, not raw speed")

Original : ['  hello ' 'WORLD' 'numpy  ' '  ARRAY']

strip    : ['hello' 'WORLD' 'numpy' 'ARRAY']
lower    : ['  hello ' 'world' 'numpy  ' '  array']
upper    : ['  HELLO ' 'WORLD' 'NUMPY  ' '  ARRAY']

strip+lower (cleaned): ['hello' 'world' 'numpy' 'array']

replace 'o'→'0': ['hell0' 'WORLD' 'numpy' 'ARRAY']

startswith 'h' (stripped): [ True False False False]
startswith 'W' (raw)     : [False  True False False]

count of 'a' per element: [0 0 0 2]
total vowels per element: [2 1 1 2]

Raw names  : ['  alice ' 'BOB' '  Carol' 'DAVE  ']
Normalized : ['Alice' 'Bob' 'Carol' 'Dave']

=== np.char vs list comprehension ===
np.char result     : ['hello' 'world' 'numpy' 'array']
listcomp result    : ['hello' 'world' 'numpy' 'array']
Results identical? : True

When to prefer each:
  np.char      → already working in NumPy pipeline, want array output directly
  listcomp     → complex logic per element (conditionals, regex, multi-step)
  Speed note   → np.char is NOT faster than listcomp (both 

### How it works

- `np.char` functions apply Python string methods **element-wise** across a string array — no explicit loop needed in your code.
- Operations can be **chained** by nesting calls: `np.char.lower(np.char.strip(arr))` strips first, then lowercases, just like calling `.strip().lower()` on a single string.
- `np.char.startswith` and `np.char.count` return **numeric or boolean arrays** — making them composable with masks and aggregations from earlier questions.
- Unlike numeric ufuncs, `np.char` functions do not use compiled vectorization internally — they loop over elements in Python. So they offer **API convenience**, not a speed advantage over list comprehensions.
- List comprehensions are preferred when logic per element is complex: multi-step conditionals, regex, or calling functions that return non-string types. `np.char` is preferred when you want to stay in the NumPy array world with clean one-liners.

**Key takeaway:** `np.char` provides convenient element-wise string operations on NumPy arrays — use it for clean pipelines in a NumPy workflow, but prefer list comprehensions when per-element logic is complex, since neither approach has a meaningful speed advantage over the other.

### Q23. `np.histogram`
Given an array of 1000 random normally distributed values, compute a histogram with 20 bins using `np.histogram`. Print the bin edges and counts. Identify the bin with the most values. Manually verify that the sum of all counts equals 1000.

In [1]:
# Q23: np.histogram — binning distributions and inspecting results

import numpy as np

rng  = np.random.default_rng(seed=42)
data = rng.normal(loc=0.0, scale=1.0, size=1000)      # 1000 values, mean=0, std=1
print("Data summary:")
print(f"  count : {len(data)}")
print(f"  mean  : {np.mean(data):.4f}  (expected ≈ 0.0)")
print(f"  std   : {np.std(data):.4f}   (expected ≈ 1.0)")
print(f"  min   : {np.min(data):.4f}")
print(f"  max   : {np.max(data):.4f}")
print()

# --- Compute histogram: 20 bins ---
counts, bin_edges = np.histogram(data, bins=20)        # returns (counts, edges)

print(f"Number of bins       : {len(counts)}")         # 20 counts
print(f"Number of bin edges  : {len(bin_edges)}")      # always counts + 1 = 21 edges
print()

# --- Print each bin with its range and count ---
print(f"{'Bin':<5} {'Range':<28} {'Count':>6}  {'Bar'}")
print("-" * 60)
for i, (count, left, right) in enumerate(zip(counts, bin_edges[:-1], bin_edges[1:])):
    bar  = "█" * (count // 10)                        # scale bar: 1 block per 10 values
    print(f"  {i:<3} [{left:+.3f}, {right:+.3f})   {count:>6}  {bar}")
print()

# --- Identify the bin with the most values ---
peak_bin   = np.argmax(counts)
peak_count = counts[peak_bin]
peak_left  = bin_edges[peak_bin]
peak_right = bin_edges[peak_bin + 1]

print(f"Peak bin      : bin {peak_bin}")
print(f"Peak range    : [{peak_left:+.4f}, {peak_right:+.4f})")
print(f"Peak count    : {peak_count} values")
print()

# --- Verify total count equals 1000 ---
total = np.sum(counts)
print(f"Sum of counts : {total}")
print(f"Matches len(data)? {total == len(data)}")
print()

# --- Custom bin edges: explicitly define boundaries ---
custom_edges  = np.array([-4, -3, -2, -1, 0, 1, 2, 3, 4], dtype=float)
counts_custom, _ = np.histogram(data, bins=custom_edges)
print("Custom bins (σ boundaries):")
print(f"{'Range':<20} {'Count':>6}  {'%':>6}")
print("-" * 36)
for cnt, lo, hi in zip(counts_custom, custom_edges[:-1], custom_edges[1:]):
    pct = cnt / len(data) * 100
    print(f"[{lo:+.0f}σ, {hi:+.0f}σ)          {cnt:>6}  {pct:>5.1f}%")
print(f"\nValues outside [-4σ, 4σ]: {len(data) - np.sum(counts_custom)}")

Data summary:
  count : 1000
  mean  : -0.0289  (expected ≈ 0.0)
  std   : 0.9887   (expected ≈ 1.0)
  min   : -3.6484
  max   : 3.1789

Number of bins       : 20
Number of bin edges  : 21

Bin   Range                         Count  Bar
------------------------------------------------------------
  0   [-3.648, -3.307)        1  
  1   [-3.307, -2.966)        0  
  2   [-2.966, -2.624)        3  
  3   [-2.624, -2.283)       10  █
  4   [-2.283, -1.942)       14  █
  5   [-1.942, -1.600)       30  ███
  6   [-1.600, -1.259)       48  ████
  7   [-1.259, -0.918)       83  ████████
  8   [-0.918, -0.576)       92  █████████
  9   [-0.576, -0.235)      126  ████████████
  10  [-0.235, +0.107)      136  █████████████
  11  [+0.107, +0.448)      160  ████████████████
  12  [+0.448, +0.789)      101  ██████████
  13  [+0.789, +1.131)       78  ███████
  14  [+1.131, +1.472)       51  █████
  15  [+1.472, +1.813)       36  ███
  16  [+1.813, +2.155)       21  ██
  17  [+2.155, +2.496)        

### How it works

- `np.histogram(data, bins=20)` divides the value range into 20 equal-width intervals and counts how many values fall into each. It returns two arrays: `counts` (length 20) and `bin_edges` (length 21 — one more than bins because each bin needs a left and right boundary).
- Each bin covers the range `[left, right)` — inclusive on the left, exclusive on the right. The last bin is `[left, right]` — inclusive on both sides.
- `np.argmax(counts)` finds the bin with the highest count — the peak of the distribution. For normally distributed data this will always be near 0 (the mean).
- `np.sum(counts)` must equal `len(data)` — every value falls into exactly one bin. This is a reliable correctness check.
- Custom bin edges let you align bins to meaningful boundaries — here, standard deviation multiples (σ). This makes it easy to verify the 68-95-99.7 rule: roughly 68% of values fall within ±1σ, 95% within ±2σ.

**Key takeaway:** `np.histogram` returns `(counts, bin_edges)` where `len(edges) = len(counts) + 1` — use `argmax(counts)` to find the peak bin, and always verify `sum(counts) == len(data)` as a correctness check.

### Q24. `np.digitize` and `np.searchsorted`
Given score data `[45, 72, 88, 55, 91, 63, 39, 78]` and grade bins `[0, 50, 65, 80, 100]`, use `np.digitize` to assign each score to a grade bin. Then use `np.searchsorted` to find where a new value `70` would be inserted into a sorted array to maintain order.

In [2]:
# Q24: np.digitize and np.searchsorted — binning and insertion points

import numpy as np

# --- np.digitize: assign each score to a grade bin ---
scores     = np.array([45, 72, 88, 55, 91, 63, 39, 78])
bins       = np.array([0, 50, 65, 80, 100])           # bin boundaries
grade_labels = ["F", "D", "C", "B", "A"]             # 5 bins → 5 labels

bin_indices = np.digitize(scores, bins, right=True)   # right=True: [lo, hi] inclusive on right

print("Scores      :", scores)
print("Bins        :", bins)
print("Bin indices :", bin_indices)                    # 1-based: 1=first bin, 2=second bin...
print()

# --- Map bin index to grade label ---
# digitize returns 1-based indices; subtract 1 to index into grade_labels
grades = np.array(grade_labels)[bin_indices - 1]
print("Grades      :", grades)
print()

# --- Detailed breakdown table ---
print(f"{'Score':>7} {'Bin idx':>9} {'Range':>16} {'Grade':>7}")
print("-" * 44)
for score, idx, grade in zip(scores, bin_indices, grades):
    lo = bins[idx - 1]
    hi = bins[idx] if idx < len(bins) else "∞"
    print(f"{score:>7}   {idx:>7}   ({lo:>3}, {hi}]   {grade:>7}")
print()

# --- right=False vs right=True: which side is inclusive ---
test_vals = np.array([50, 65, 80])                    # values exactly on boundaries
print("Boundary behaviour for values exactly on bin edges:")
print("right=False (default) — [lo, hi):", np.digitize(test_vals, bins, right=False))
print("right=True            — (lo, hi]:", np.digitize(test_vals, bins, right=True))
print("  right=False: 50 goes INTO bin starting at 50 (left-inclusive)")
print("  right=True : 50 goes INTO bin ending   at 50 (right-inclusive)")
print()

# --- np.searchsorted: find insertion point to maintain sorted order ---
sorted_arr = np.array([10, 20, 30, 40, 50, 60, 70, 80, 90])
new_val    = 70

idx_left  = np.searchsorted(sorted_arr, new_val, side='left')   # leftmost valid position
idx_right = np.searchsorted(sorted_arr, new_val, side='right')  # rightmost valid position

print("Sorted array   :", sorted_arr)
print(f"Insert value   : {new_val}")
print(f"side='left'    : index {idx_left}  → insert BEFORE existing {new_val}")
print(f"side='right'   : index {idx_right}  → insert AFTER  existing {new_val}")
print()

# Verify by inserting
inserted_left  = np.insert(sorted_arr, idx_left,  new_val)
inserted_right = np.insert(sorted_arr, idx_right, new_val)
print("After insert (left) :", inserted_left)
print("After insert (right):", inserted_right)
print()

# --- Practical: searchsorted on multiple values at once ---
new_scores  = np.array([25, 55, 73, 95])
positions   = np.searchsorted(sorted_arr, new_scores, side='left')
print("New scores      :", new_scores)
print("Insert positions:", positions)
for val, pos in zip(new_scores, positions):
    print(f"  {val} would be inserted at index {pos}"
          f"  (between {sorted_arr[pos-1]} and {sorted_arr[pos] if pos < len(sorted_arr) else '∞'})")

Scores      : [45 72 88 55 91 63 39 78]
Bins        : [  0  50  65  80 100]
Bin indices : [1 3 4 2 4 2 1 3]

Grades      : ['F' 'C' 'B' 'D' 'B' 'D' 'F' 'C']

  Score   Bin idx            Range   Grade
--------------------------------------------
     45         1   (  0, 50]         F
     72         3   ( 65, 80]         C
     88         4   ( 80, 100]         B
     55         2   ( 50, 65]         D
     91         4   ( 80, 100]         B
     63         2   ( 50, 65]         D
     39         1   (  0, 50]         F
     78         3   ( 65, 80]         C

Boundary behaviour for values exactly on bin edges:
right=False (default) — [lo, hi): [2 3 4]
right=True            — (lo, hi]: [1 2 3]
  right=False: 50 goes INTO bin starting at 50 (left-inclusive)
  right=True : 50 goes INTO bin ending   at 50 (right-inclusive)

Sorted array   : [10 20 30 40 50 60 70 80 90]
Insert value   : 70
side='left'    : index 6  → insert BEFORE existing 70
side='right'   : index 7  → insert AFTER  exi

### How it works

- `np.digitize(values, bins)` assigns each value to a bin and returns a **1-based index** — bin 1 is between `bins[0]` and `bins[1]`, bin 2 between `bins[1]` and `bins[2]`, and so on. Subtract 1 before using as a list index.
- `right=False` (default) makes each bin left-inclusive `[lo, hi)` — a value exactly on the boundary goes into the next bin. `right=True` makes bins right-inclusive `(lo, hi]` — boundary values belong to the current bin. For grades, `right=True` is natural: a score of exactly 80 should earn a B, not a C.
- `np.searchsorted(arr, val)` finds the index where `val` should be inserted to keep the array sorted — using binary search, so it is very fast even on large arrays.
- `side='left'` inserts **before** any existing equal values; `side='right'` inserts **after**. This matters when duplicates are present.
- Both functions require the bins/array to be **sorted in ascending order** — results are undefined otherwise.

**Key takeaway:** `np.digitize` maps continuous values to labelled bins (like grade categories); `np.searchsorted` finds the exact insertion point to maintain sorted order — both use binary search internally and work on entire arrays at once.

### Q25. Challenge — Vectorized Statistics Without Loops
Given a 2D dataset of shape `(100, 5)` (100 observations, 5 features) filled with random data, compute the following **without any Python loops**:
- Mean and std of each feature (column)
- Z-score normalization of the entire dataset: `(X - mean) / std`
- Number of outliers per feature (values more than 2 std away from mean)
- The feature (column index) with the highest variance
- The row index of the observation closest to the mean vector (using Euclidean distance, fully vectorized)
Print all results with clear labels.
 

In [3]:
# Q25: Vectorized statistics on a 2D dataset — no Python loops

import numpy as np

rng  = np.random.default_rng(seed=0)
X    = rng.normal(loc=0.0, scale=1.0, size=(100, 5))  # shape (100, 5)
print("Dataset shape:", X.shape)
print()

# --- 1. Mean and std of each feature (column) ---
col_mean = np.mean(X, axis=0)                          # shape (5,) — one value per column
col_std  = np.std(X,  axis=0)                          # shape (5,)

print("=== Feature statistics ===")
print(f"{'Feature':<10} {'Mean':>10} {'Std':>10}")
print("-" * 32)
for i, (m, s) in enumerate(zip(col_mean, col_std)):
    print(f"  F{i:<7}  {m:>+10.4f}  {s:>10.4f}")
print()

# --- 2. Z-score normalization: (X - mean) / std ---
# keepdims=True keeps shape (1,5) so broadcasting subtracts per column correctly
Z = (X - col_mean) / col_std                           # broadcasting: (100,5) - (5,) → (100,5)
print("=== Z-score normalization ===")
print("Z shape          :", Z.shape)
print("Z column means   :", np.round(np.mean(Z, axis=0), 6))   # should be ≈ 0.0
print("Z column stds    :", np.round(np.std(Z,  axis=0), 6))   # should be ≈ 1.0
print()

# --- 3. Number of outliers per feature (|z| > 2) ---
outlier_mask  = np.abs(Z) > 2                          # shape (100, 5) — True where outlier
outlier_count = np.sum(outlier_mask, axis=0)           # shape (5,) — count per column

print("=== Outliers per feature (|z| > 2) ===")
for i, (cnt, pct) in enumerate(zip(outlier_count, outlier_count / 100 * 100)):
    print(f"  F{i}: {cnt:>3} outliers  ({pct:.1f}%)  expected ≈ 5% → ~5 values")
print(f"Total outliers across all features: {np.sum(outlier_mask)}")
print()

# --- 4. Feature with highest variance ---
col_var      = np.var(X, axis=0)                       # shape (5,)
highest_var  = np.argmax(col_var)                      # index of max variance

print("=== Variance per feature ===")
for i, v in enumerate(col_var):
    marker = " ◄ highest" if i == highest_var else ""
    print(f"  F{i}: var = {v:.4f}{marker}")
print(f"Feature with highest variance: F{highest_var}")
print()

# --- 5. Row closest to the mean vector (Euclidean distance, fully vectorized) ---
# mean vector has shape (5,); X has shape (100, 5)
# subtract broadcasts: (100,5) - (5,) → diff shape (100,5)
diff      = X - col_mean                               # deviation from mean per element
dist      = np.sqrt(np.sum(diff ** 2, axis=1))        # Euclidean distance per row, shape (100,)
closest   = np.argmin(dist)                            # row index of smallest distance

print("=== Observation closest to mean vector ===")
print(f"Mean vector        : {np.round(col_mean, 4)}")
print(f"Closest row index  : {closest}")
print(f"Closest row values : {np.round(X[closest], 4)}")
print(f"Euclidean distance : {dist[closest]:.4f}")
print()

# Sanity check: confirm no row is closer
print(f"Min distance in dist array     : {np.min(dist):.4f}")
print(f"dist[{closest}] == np.min(dist)       :", np.isclose(dist[closest], np.min(dist)))

Dataset shape: (100, 5)

=== Feature statistics ===
Feature          Mean        Std
--------------------------------
  F0           -0.0893      0.9039
  F1           -0.1025      0.9954
  F2           +0.1134      0.8888
  F3           -0.2123      1.1369
  F4           +0.1563      1.0721

=== Z-score normalization ===
Z shape          : (100, 5)
Z column means   : [-0.  0.  0. -0. -0.]
Z column stds    : [1. 1. 1. 1. 1.]

=== Outliers per feature (|z| > 2) ===
  F0:   4 outliers  (4.0%)  expected ≈ 5% → ~5 values
  F1:   3 outliers  (3.0%)  expected ≈ 5% → ~5 values
  F2:   3 outliers  (3.0%)  expected ≈ 5% → ~5 values
  F3:   5 outliers  (5.0%)  expected ≈ 5% → ~5 values
  F4:   5 outliers  (5.0%)  expected ≈ 5% → ~5 values
Total outliers across all features: 20

=== Variance per feature ===
  F0: var = 0.8171
  F1: var = 0.9907
  F2: var = 0.7900
  F3: var = 1.2925 ◄ highest
  F4: var = 1.1494
Feature with highest variance: F3

=== Observation closest to mean vector ===
Mean vect

### How it works

- `np.mean(X, axis=0)` and `np.std(X, axis=0)` collapse the 100-row dimension, returning shape `(5,)` — one statistic per feature column. No loop over columns needed.
- Z-score normalization `(X - col_mean) / col_std` uses broadcasting: `col_mean` has shape `(5,)` which NumPy aligns to `(100, 5)` automatically — subtracting the correct mean from every row simultaneously. After normalization, every column has mean ≈ 0 and std ≈ 1.
- Outlier detection uses `np.abs(Z) > 2` — a single comparison producing a boolean `(100, 5)` mask. `np.sum(mask, axis=0)` counts `True` values per column. For a normal distribution, roughly 5% of values (≈5 out of 100) are expected beyond ±2σ.
- `np.var(X, axis=0)` gives variance per feature; `np.argmax` on that returns the column index with the highest spread — all in two lines.
- Euclidean distance is computed fully vectorized: `diff = X - col_mean` gives per-element deviations `(100, 5)`, `diff**2` squares them, `np.sum(..., axis=1)` sums across features giving shape `(100,)`, and `np.sqrt` gives the final distances. `np.argmin` finds the closest row.

**Key takeaway:** All five statistical tasks — column stats, normalization, outlier detection, variance ranking, and nearest-row search — are solved purely with broadcasting, axis-based aggregation, and boolean masks, with zero Python loops and in just a few lines each.